hs_core/management/commands/migrate_script_resources.py

```diff
@@ -0,0 +1,82 @@
+from django.core.management.base import BaseCommand
+
+from hs_core.hydroshare import current_site_url, set_dirty_bag_flag
+from hs_core.models import CoreMetaData
+from hs_script_resource.models import ScriptResource
+from ..utils import migrate_core_meta_elements
+
+
+class Command(BaseCommand):
+    help = "Convert all script resources to composite resource"
+
+    def handle(self, *args, **options):
+        resource_counter = 0
+        to_resource_type = 'CompositeResource'
+        msg = "THERE ARE CURRENTLY {} SCRIPT RESOURCES TO MIGRATE TO COMPOSITE RESOURCE.".format(
+            ScriptResource.objects.count())
+        print(msg, flush=True)
+        for script_res in ScriptResource.objects.all():
+            msg = "Migrating script resource:{}".format(script_res.short_id)
+            print(msg, flush=True)
+            # check resource exists on irods
+            istorage = script_res.get_irods_storage()
+            if not istorage.exists(script_res.root_path):
+                err_msg = "Script resource was not found in iRODS (ID: {})".format(script_res.short_id)
+                print("ERROR: {}".format(err_msg), flush=True)
+                # skip this script resource for migration
+                continue
+
+            # change the resource_type
+            script_metadata_obj = script_res.metadata
+            script_res.resource_type = to_resource_type
+            script_res.content_model = to_resource_type.lower()
+            script_res.save()
+
+            # get the converted resource object - CompositeResource
+            comp_res = script_res.get_content_model()
+
+            # set CoreMetaData object for the composite resource
+            core_meta_obj = CoreMetaData.objects.create()
+            comp_res.content_object = core_meta_obj
+            # migrate script resource core metadata elements to composite resource
+            migrate_core_meta_elements(script_metadata_obj, comp_res)
+            # dtarb suggested to add this extended metadata
+            comp_res.extra_metadata['Type'] = 'Script'
```

Comment:
Is there a schema documented for the organization of Script in extra_metadata?  

It is a cumbersome system to read/write custom metadata this way.  It is stored in the rdf/xml as a list of key/value pairs.  If a user wants to interact with it, they would have to know to parse a list of key/value pairs and load it into a dictionary.  They would have to do the reverse to update the resource.

How do the users know to look at extra_metadata and parse out the values?

I'm afraid that we are making things more difficult for ourselves in the long run here.  What if instead, we allowed users to extend our schemas?

Reply:
I am not sure how the existing script resource type has been used by the user. My understanding of this migration is to get rid of the script resource type and preserve the resource files and metadata as part of the composite resource. If we need well defined script specific metadata, I assume we have to support a new aggregation type. May be @dtarb can comment on this. 

> What if instead, we allowed users to extend our schemas?

Which schema? May be I am missing something.

Reply:
What I'm proposing is certainly more work than just this schema within extended metadata but it will be more useful in the long run.  This is probably not the place to have the conversation and I guess I'm thinking about the work with hsmodels.

We have json schemas of our resource metadata and our aggregation metadata.  With the Submission page that the Data Submission Portal introduces, we can take any (most?) json schemas and generate forms for it.  We could allow other users to specify their own schemas while also validating for the properties we expect in our schemas defined in hsmodels.  This would be a solution to model program/instance as well.

I was originally thinking it'd be nice to allow users to specify a schema on extended metadata similar to how we allow users to specify a schema with model program.  But there is really no good reason to limit it to extended metadata.  

In short, I'm fine with this change because it solves the problem of dissolving script resource for the small handful of instances we have in HydroShare.  It just made me think of how to solve this problem more broadly.

Reply:
I assume hsmodels currently only supports predefined schemas. That's why I was not sure how the user extended schema would work. I agree as a general solution we need to support user extensible schemas as part of hsmodels.